fix(ngcc): correctly get config for packages in nested `node_modules/`

gkalpak opened the pull request:
Previously, ngcc would only be able to match an ngcc configuration to packages that were located inside the project's top-level `node_modules/`. However, if there are multiple versions of a package in a project (e.g. as a transitive dependency of other packages), multiple copies of a package (at different versions) may exist in nested `node_modules/` directories. For example, one at `<project-root>/node_modules/some-package/` and one at `<project-root>/node_modules/other-package/node_modules/some-package/`. In such cases, ngcc was only able to detect the config for the first copy but not for the second.

This commit fixes this by returning a new instance of `ProcessedNgccPackageConfig` for each different package path (even if they refer to the same package name). In these `ProcessedNgccPackageConfig`, the `entryPoints` paths have been processed to take the package path into account.

mary-poppins commented:
You can preview ac92b5a at https://pr37040-ac92b5a.ngbuilds.io/.

devversion reviewed (approved):
LGTM for dev-infra.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/configuration.ts:
```diff
@@ -44,17 +44,16 @@ export interface ProcessLockingConfiguration {
+   * In the config file the keys can be paths relative to the package path.
    */
-  entryPoints: {[entryPointPath: string]: NgccEntryPointConfig;};
+  entryPoints?: {[entryPointPath: string]: NgccEntryPointConfig|undefined};
```
```suggestion
  entryPoints?: {[entryPointPath: string]: NgccEntryPointConfig};
```

gkalpak replied:
The reason I added `undefined` is that it will better reflect the fact that `.entryPoints.someEntryPoint` might or might not be defined. Previously, TS would infer it as `NgccEntryPointConfig` although that was not necessarily the case.

This was more of an issue when this type was returned by `getPackageConfig()`, so happy to remove it now that `RawNgccPackageConfig` is only used internally.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/configuration.ts:
```diff
@@ -152,13 +158,48 @@ export const DEFAULT_NGCC_CONFIG: NgccProjectConfig = {
+    const absolutePathEntries: [AbsoluteFsPath, NgccEntryPointConfig][] =
+        Object.entries(entryPoints).map(([
+                                          relativePath, config
+                                        ]) => [resolve(packagePath, relativePath), config!]);
```
If this is the case, then it is better to remove the `undefined` from the definition of:

```
entryPoints?: {[entryPointPath: string]: NgccEntryPointConfig|undefined};
```

gkalpak replied:
The main purpose of `undefined` was to guard against the fact that accessing arbitrary properties on `entryPoints` is not guaranteed to return an `NgccEntryPointConfig` (see https://github.com/angular/angular/pull/37040#discussion_r424670933). Here we are accessing properties that we know exist on `entryPoints`, so it is safe(r) to assume they are not `undefined`. (Not that it is not possible for people to set `entryPoints.foo` to undefined in their configs, but addressing that would be outside the scope of this PR :grin:)

Removing the `undefined` as mentioned in https://github.com/angular/angular/pull/37040#discussion_r424670933...

petebacondarwin reviewed (changes requested):
Can you remove the `undefined` from the `RawNgccPackageConfig.exports` declaration.
Otherwise LGTM (with a nit or two).

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -118,26 +118,28 @@ export type GetEntryPointResult = EntryPoint|typeof INCOMPATIBLE_ENTRY_POINT|typ
 export function getEntryPointInfo(
     fs: FileSystem, config: NgccConfiguration, logger: Logger, packagePath: AbsoluteFsPath,
     entryPointPath: AbsoluteFsPath): GetEntryPointResult {
+  const packageName = packagePath.replace(/^.*\/node_modules\//i, '');
```
I thought about it too, but I felt it would require `getPackageConfig()` to assume too much. It is an implementation detail that `getEntryPointInfo()` only considers packages inside `node_modules/` (which might change in the future) and I think it is the caller's responsibility to determine the package name.

The dependency resolver, for example, does it differently: https://github.com/angular/angular/blob/ac92b5ac6c79f0a7cc2900b48bf8f6a3f5a85aaf/packages/compiler-cli/ngcc/src/dependencies/dependency_resolver.ts#L228-L229

WDYT?

gkalpak commented:
Addressed feedback (or way or another 😛). @petebacondarwin, ptal.

mary-poppins commented:
You can preview cae6180 at https://pr37040-cae6180.ngbuilds.io/.

mary-poppins commented:
You can preview 577858a at https://pr37040-577858a.ngbuilds.io/.

mary-poppins commented:
You can preview dc697b0 at https://pr37040-dc697b0.ngbuilds.io/.

mary-poppins commented:
You can preview bd5b369 at https://pr37040-bd5b369.ngbuilds.io/.

mary-poppins commented:
You can preview 2f1b8c8 at https://pr37040-2f1b8c8.ngbuilds.io/.

gkalpak commented:
This turned out to be a little more tricky than I expected. I finally have something I am happy with (I think 😅).

@petebacondarwin, ptal.

petebacondarwin commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -294,16 +301,41 @@ function guessTypingsFromPackageJson(
+
+  return {
+    packageName,
+    packageVersion: packageJson?.version ?? null,
```
If the entry-point package.json does not contain a `version` property, which is possible, then we could just load up the `package.json` of the main package (found at `packagePath`)?

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts:
```diff
@@ -166,6 +169,136 @@ runInEachFileSystem(() => {
+            const entryPoint = getEntryPointInfo(fs, config, logger, packagePath, entryPointPath);
+
+            // Ensure we did not get `NO_ENTRY_POINT` or `INCOMPATIBLE_ENTRY_POINT`.
+            expect(entryPoint).toBeInstanceOf(Object);
```
Another opportunity to use the `isEntryPoint()` helper suggested earlier.

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts:
```diff
@@ -166,6 +169,136 @@ runInEachFileSystem(() => {
+              }
+
+              // Ensure there is an ngcc config for the entry-point.
+              getPackageConfigSpy.withArgs(packagePath, null).and.returnValue({
```
Could we avoid a spy and just generate an `ngcc.config.js` file here?

gkalpak replied:
I guess we could, but what would be the benefit?

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts:
```diff
@@ -166,6 +169,136 @@ runInEachFileSystem(() => {
+                },
+              ]);
+            } else {
+              // Ensure the entry-point `package.json` does not exist (e.g. from a previous test).
```
Do you mean from a previous `expect()`? Since each `it()` clause should get its own new `MockFileSystem`.
Perhaps it would be safer to move each of these `expect()` calls into their own `it()` clause if we are concerned about bleeding data between them?

petebacondarwin replied:
This would also allow us to have a description for each of them, which would be good documentation of what the tests are actually doing.

gkalpak replied:
Yes, from a previous `expect()`. I thought about splitting up into different `it()` calls, but I felt it wouldn't make the tests easier to follow/read.

gkalpak replied:
Happy to split it up if you think it would be better.

petebacondarwin reviewed (changes requested):
I am not sure how it passes integration tests but I am pretty sure we cannot rely upon secondary entry-points containing a `version` property.

gkalpak commented on packages/compiler-cli/ngcc/src/entry_point_finder/directory_walker_entry_point_finder.ts:
```diff
@@ -148,14 +150,15 @@ export class DirectoryWalkerEntryPointFinder implements EntryPointFinder {
+      const isEntryPoint =
+          subEntryPoint !== NO_ENTRY_POINT && subEntryPoint !== INCOMPATIBLE_ENTRY_POINT;
+      if (subEntryPoint !== NO_ENTRY_POINT && subEntryPoint !== INCOMPATIBLE_ENTRY_POINT &&
+          subEntryPoint !== IGNORED_ENTRY_POINT) {
```
It felt less readable. How about re-using `isEntryPoint`:

```ts
const isEntryPoint =
    subEntryPoint !== NO_ENTRY_POINT && subEntryPoint !== INCOMPATIBLE_ENTRY_POINT;
if (isEntryPoint && subEntryPoint !== IGNORED_ENTRY_POINT) {
```

mary-poppins commented:
You can preview 71ae344 at https://pr37040-71ae344.ngbuilds.io/.

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -294,16 +301,41 @@ function guessTypingsFromPackageJson(
+ *   from the package's path.
+ * - The version is read off of the `version` property of `packageJson`. It is assumed that even if
+ *   `packageJson` corresponds to a secondary entry-point (e.g. `@angular/common/http`) it will
+ *   still contain the same version as the main package (e.g. `@angular/common`).
```
😱

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -299,16 +296,14 @@ function guessTypingsFromPackageJson(
- * @returns the version string or `null` if the package.json does not exist or is invalid.
+ * The version is read off of the `version` property of `packageJson`. It is assumed that even if
+ * `packageJson` corresponds to a secondary entry-point (e.g. `@angular/common/http`) it will still
+ * contain the same version as the main package (e.g. `@angular/common`).
```
😱

gkalpak commented on packages/compiler-cli/ngcc/src/packages/entry_point.ts:
```diff
@@ -124,31 +125,32 @@ export function getEntryPointInfo(
-    // package.json exists but could not be parsed and there was no redeeming config
-    return INCOMPATIBLE_ENTRY_POINT;
-  }
+      loadedEntryPointPackageJson!;
```
Agreed this is not great. I have moved things around to get rid of the non-null assertions.

mary-poppins commented:
You can preview 82ce6c7 at https://pr37040-82ce6c7.ngbuilds.io/.

mary-poppins commented:
You can preview 45318c8 at https://pr37040-45318c8.ngbuilds.io/.

mary-poppins commented:
You can preview c882ee1 at https://pr37040-c882ee1.ngbuilds.io/.

gkalpak committed:
refactor(ngcc): fix typos in comments

This is a follow-up to #36944, because I didn't manage to finish my
review before the PR got merged.

gkalpak committed:
refactor(ngcc): clean up unused imports, unused regex parenthesis, typos

This is a follow-up to #37075, because I didn't manage to finish my
review before the PR got merged.

gkalpak committed:
fix(ngcc): correctly get config for sub-entry-points when primary entry-point is ignored

Previously, when an entry-point was ignored via an ngcc config, ngcc
would scan sub-directories for sub-entry-points, but would not use the
correct `packagePath`. For example, if `@angular/common` was ignored, it
would look at `@angular/common/http` but incorrectly use
`.../@angular/common/http` as the `packagePath` (instead of
`.../@angular/common`). As a result, it would not retrieve the correct
ngcc config for the actual package.

This commit fixes it by ensuring the correct `packagePath` is used, even
if the primary entry-point corresponding to that path is ignored. In
order to do this, a new return value for `getEntryPointInfo()` is added:
`IGNORED_ENTRY_POINT`. This is used to differentiate between directories
that correspond to no or an incompatible entry-point and those that
correspond to an entry-point that could otherwise be valid but is
explicitly ignored. Consumers of `getEntryPointInfo()` can then use this
info to discard ignored entry-points, but still use the correct
`packagePath` when scanning their sub-directories for secondary
entry-points.

gkalpak committed:
refactor(ngcc): rename `EntryPoint#package` to `EntryPoint#packagePath`

Rename the `package` property to `packagePath` on the `EntryPoint`
interface. This makes it more clear that the `packagePath` property
holds the absolute path to the containing package (similar to how `path`
holds the path to the entry-point). This will also align with the
`packageName` property that will be added in a subsequent commit.

This commit also re-orders the `EntryPoint` properties to group related
properties together and to match the order of properties on instances
with that on the interface.

gkalpak committed:
refactor(ngcc): refactor how info is retrieved from entry-point `package.json`

This commit refactors the way info is retrieved from entry-point
`package.json` files to make it easier to extract more info (such as the
package's name) in the future. It also avoids reading and parsing the
`package.json` file multiple times (as was happening before).

gkalpak committed:
fix(ngcc): correctly retrieve a package's version from its `package.json`

In order to retrieve the ngcc configuration (if any) for an entry-point,
ngcc has to detect the containing package's version.

Previously, ngcc would try to read the version from the entry-point's
`package.json` file, which was different than the package's top-level
`package.json` for secondary entry-points. For example, it would try to
read it from `node_modules/@angular/common/http/package.json` for
entry-point `@angular/common/http`. However, the `package.json` files
for secondary entry-points are not guaranteed to include a `version`
property.

This commit fixes this by first trying to read the version from the
_package's_ `package.json` (falling back to the entry-point's
`package.json`). For example, it will first try to read it from
`@angular/common/package.json` for entry-point `@angular/common/http`.

gkalpak committed:
refactor(ngcc): add `packageName` property to `EntryPoint` interface

This commit adds a `packageName` property to the `EntryPoint` interface.
In a subsequent commit this will be used to retrieve the correct ngcc
configuration for each package, regardless of its path.

gkalpak committed:
fix(ngcc): correctly get config for packages in nested `node_modules/`

Previously, ngcc would only be able to match an ngcc configuration to
packages that were located inside the project's top-level
`node_modules/`. However, if there are multiple versions of a package in
a project (e.g. as a transitive dependency of other packages), multiple
copies of a package (at different versions) may exist in nested
`node_modules/` directories. For example, one at
`<project-root>/node_modules/some-package/` and one at
`<project-root>/node_modules/other-package/node_modules/some-package/`.
In such cases, ngcc was only able to detect the config for the first
copy but not for the second.

This commit fixes this by returning a new instance of
`ProcessedNgccPackageConfig` for each different package path (even if
they refer to the same package name). In these
`ProcessedNgccPackageConfig`, the `entryPoints` paths have been
processed to take the package path into account.

gkalpak commented:
@petebacondarwin, I have addressed your comments in fixup commits (all fixup commits are new) plus this new regular commit:
- **fix(ngcc): correctly retrieve a package's version from it's `package.json`** (https://github.com/angular/angular/pull/37040/commits/8d353e36fd0db65bf8830a8418291c46afffa8ae)

mary-poppins commented:
You can preview e9a8e1b at https://pr37040-e9a8e1b.ngbuilds.io/.

petebacondarwin reviewed (approved):
LGTM - Just the one question about the order of checking for entry-point version that I think is worth considering.

gkalpak committed:
fixup! fix(ngcc): correctly retrieve a package's version from its `package.json`

gkalpak committed:
fixup! refactor(ngcc): add `packageName` property to `EntryPoint` interface

gkalpak committed:
fixup! fix(ngcc): correctly retrieve a package's version from its `package.json`

petebacondarwin reviewed (approved):
LGTM (again)

JoostK reviewed (approved):
Nice cleanup overall!
